Simplify fluxyaml reference (#2634)
fluxcd/flux: docs/references/fluxyaml-config-files.md

# Manifest generation through `.flux.yaml` configuration files

This feature lets you generate Kubernetes manifests with a program,

> instead of having to include them in your git repo as YAML files. For

Reviewer: ```suggestion
instead of having to include them in your Git repository as YAML files. For
```

I am not sure we are consistent in terminology across the docs. If we use _git_ and _repo_ everywhere else I am fine with it. `Consistency > formal pickiness`

Author: We use "git repo" and "git repository" here and there. Of those I think I prefer "git repository". Using a capital gives it a bit too much emphasis I reckon.

Author: I've used "git repository" consistently, now.

example, you can use `kustomize` to patch a common set of resources to
suit a particular environment.

## Setting manifest generation up

Manifest generation is controlled by the flags given to `fluxd`, and

> `.flux.yaml` files in your git repo.

Reviewer: Same

> To enable it, pass the command-line flag `--manifest-generation=true`

Reviewer: Maybe we should highlight this a bit more, since it's essential for the feature. Do we use a highlight box or something like that in other parts of the docs?

Author: I moved it to the top and made the absolute requirements clearer.

to `fluxd`.

> For each path supplied with the flag `--git-path` (or `.` if none are

Reviewer: Maybe we should introduce what `--git-path` is beforehand.

Reviewer: Also, it may be worth clarifying that `.` means the root directory of the repository.

Reviewer: And, I am surely biased but, wouldn't it make sense to (at least briefly) present a `.flux.yaml` file before getting into their interaction with `--git-path`?

As a reader I am not sure I would be able to understand what is going on without having a rough high-level idea of what the configuration will do on that `--git-path`

Reviewer: > And, I am surely biased but, wouldn't it make sense to (at least briefly) present a `.flux.yaml` file before getting into their interaction with `--git-path`?

I think it would make sense to show it and indicate it will be explained later.

Reviewer 2: I agree that explaining what `--git-path` does would make the docs better.

Author: Cool, OK. I was self-consciously not piling up too much text before getting to the good stuff, but you are both right that `--git-path` deserves a bit of explaining.

Author: I put a bit of explanation up-front, and substituted the term "target path" for `--git-path` afterwards. I got the term from the earlier text, and it seems OK to me, but I'm happy to adopt another.

supplied),

 - if a `.flux.yaml` file is found, it is used **instead** of looking
   for YAML files;

>  - if no `.flux.yaml` file is found, the usual behaviour of looking

Reviewer: It may be worth briefly describing the usual behavior

Author: "looking for YAML files" pretty much is a brief explanation of the usual behaviour, isn't it?

Reviewer 3: How about "the usual behaviour of recursively reading all YAML files under the target path."

   for YAML files is used.

The manifests from all the `--git-path` values -- read from YAML files
or generated -- are combined before applying to the cluster. If

> duplicates are detected, an error is logged and the sync is abandoned.

Reviewer: ```suggestion
duplicates are detected, an error is logged and Flux will abort the synchronization with the cluster.
```

Author: Rephrased, though not exactly as suggested

A `.flux.yaml` file can be _at_ the path mentioned, or in a directory
_above_ it in the git repo. Here are some examples:

```sh
# show normal files _and_ .flux.yaml

> flux-kustomization-example$ tree -a -P .flux.yaml -P '*' -I .git

Reviewer: Is the _sophisticated_ tree command and `flux-kustomization-example` name necessary? I think it doesn't add much value as is and can add confusion.

I think we should either be explicit about using a real-world example (even adding a prior `git clone` command on https://github.com/weaveworks/flux-kustomize-example ) or just show the tree.

Reviewer 2: I would go with a `git clone && cd flux-kustomization-example` command and remove `flux-kustomization-example$`.

Reviewer: BTW, I should move the repo to the `fluxcd` org.

@squaremo Do you want me to do it before merging this PR?

Reviewer 2: GitHub deals with redirects, I would just move it now.

Author: I wanted to show a directory structure, but it doesn't really matter how I get it, so I'll just remove the command from the examples.

.
├── base
│   ├── demo-ns.yaml
│   ├── kustomization.yaml
│   ├── podinfo-dep.yaml
│   ├── podinfo-hpa.yaml
│   └── podinfo-svc.yaml
├── .flux.yaml
├── LICENSE
├── production
│   ├── flux-patch.yaml
│   ├── kustomization.yaml
│   └── replicas-patch.yaml
├── README.md
└── staging
    ├── flux-patch.yaml
    └── kustomization.yaml
```

In this case, say you started `fluxd` with `--git-path=staging`, it
would find `.flux.yaml` in the top directory and use that to generate
manifests.  The files and directories (if there were any) in

> `staging/` are ignored, in favour of the instructions given in the

Reviewer: Nitpick (the files are actually used)
```suggestion
`staging/` are not traversed by Flux, in favour of the instructions given in the
```

Author: Rephrased, with "examined" rather than "traversed"

`.flux.yaml` file.

> In this modified example,

Reviewer: I think it's worth indicating what is being changed (I had to look at it a couple of times).

Author: Yes good point that. Also adopted your suggestion below of sticking with `staging` in the example.

```sh
flux-kustomization-example$ tree -a -P .flux.yaml -P '*' -I .git
.
├── base
│   ├── demo-ns.yaml
│   ├── kustomization.yaml
│   ├── podinfo-dep.yaml
│   ├── podinfo-hpa.yaml
│   └── podinfo-svc.yaml
├── LICENSE
├── production
│   ├── flux-patch.yaml
│   ├── .flux.yaml
│   ├── kustomization.yaml
│   └── replicas-patch.yaml
├── README.md
└── staging
    ├── flux-patch.yaml
    └── kustomization.yaml
```

> … the `.flux.yaml` file is under `production/`, so it will _only_

Reviewer: Super nitpick: I would probably change the example to moving the `.flux.yaml` file inside `staging` and saying that it's equivalent to the prior configuration, after setting `--git-path=staging`.

Also, it may be worth mentioning that (before making the change) you can specify the production environment by specifying `--git-path=production` without moving the `.flux.yaml` file.

take effect for `--git-path=production`.

Using `--git-path=staging` would **not produce a usable
configuration**, because without an applicable `.flux.yaml`, the files
under `staging/` would be treated as plain Kubernetes manifests, which
they are plainly not.

Note also that the configuration file would **not** take effect for
`--git-path=.` (i.e., the top directory), because manifest generation
will not look in subdirectories for a `.flux.yaml` file.

## How to construct a .flux.yaml file

`.flux.yaml` files come in two varieties: "patch-updated", and
"command-updated". These refer to the way in which automated updates

> are applied to files in the repo:

Reviewer: Worth adding a link to the relevant documentation for automated updates?

programs to find and update the right file, it is likely to be a lot
more work.

Both patch-updated and command-updated configurations have the same
way of specifying how to generate manifests, and differ only in how
updates are recorded.

### Generator configuration

Here is an example of a `.flux.yaml`:

```yaml
version: 1 # must be `1`
patchUpdated:
  generators:
  - command: kustomize build .
  patchFile: flux-patch.yaml
```

The `generators` field is an array of commands, all of which will be
run in turn. Each command is expected to print a YAML stream to its
stdout. The streams are concatenated and parsed as one big YAML
stream, before being applied.

Much of the time, it will only be necessary to supply one command to
be run.

The commands will be run with the `--git-path` being processed as a
working directory -- which is not necessarily the same directory in
which the `.flux.yaml` file was found. [See below](#execution-context)
for more details on the execution context in which commands are run.

### Using patch-updated configuration

A patch-updated configuration generates manifests using commands, and

> records updates in a patch file.

Reviewer: Is it worth mentioning it's a Strategic Merge patch? We could hide that fact to avoid confusion, but it's their responsibility consistent with the manifests they generate.

For example, when an automated image upgrade is run, fluxd will do
this:

 1. run the generator commands and parse the manifests;
 2. find the manifest that needs to be updated, and calculate the
    patch to it that performs the update;
 3. record that patch in the patch file.

When syncing, fluxd will generate the manifests as usual, then apply
all the patches that have been recorded in the patch file.

This is how a patch-updated `.flux.yaml` looks:

```yaml
version: 1
patchUpdated:
  generators:
  - command: jk generate ./index.js
  patchFile: flux-patch.yaml
```

The `generators` field is explained just above. The `patchFile` field
gives a path, relative to the `--git-path` value, in which to record

> patches. `fluxd` will create the file if necessary, and commit any

Reviewer: Nitpick:
```suggestion
patches. `fluxd` will create the file when needed, and commit any
```

Reviewer: Also, it may be worth mentioning that it's the user's responsibility to keep the patch file consistent with the generated manifests (e.g. if a manifests is removed, their patch counterpart needs to be removed too).

Reviewer 2: Many users were confused about GC and patches. We should mention that even if Flux GC removes a deployment the patch will error out if not removed manually

Author: I've mentioned the need to manually remove patches when manifests go.

commandUpdated:
  generators:
  - command: generator_command
  updaters:
  - containerImage:
      command: image_updater_program
    policy:
      command: policy_updater_program
```

The `updaters` section is particular to command-updated
configuration. It contains an array of updaters, each of which gives a
command for updating container images, and a command for updating
policies (policy controls how automated updates should be applied to a
resource; these appear as annotations in generated manifests).

When asked to update a resource, fluxd will run execute the
appropriate variety of command for *each* entry in `updaters:`. For
example, when updating an image, it will execute the command under
`containerImage`, for each updater entry, in turn.

Usually updates come in batches -- e.g., updating the same container
image in several resources -- so the commands will likely be run
several times.

### Execution context of commands

`generators` and `updaters` are run in a POSIX shell inside the fluxd
container. This means that the executables mentioned in commands must
be available in the [Flux container
image](https://github.com/fluxcd/flux/blob/master/docker/Dockerfile.flux).

> Flux currently includes `kustomize` and basic Unix shell tools. If the

Reviewer: ```suggestion
Flux currently includes `kustomize`, `helm` and basic Unix shell tools. If the
```

Reviewer: Ignore this, we don't include the binary in the image (we only use it for testing).

tools in the Flux image are not sufficient for your use case, you can

> include new tools in your own Flux-based image or, if the tools are

Reviewer: Is it worth mentioning that an `init` container can be used, without needing to create a custom image?

e.g. https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-init-containers-plugin-downloads.html

Maybe we can include an example adding an extra binary.

Author: Umm yes it probably is worth giving an example of that, fiddly though it is. Maybe I'll do so in another commit.

Author: I've mentioned it, along with building a custom image, as one of the possibilities here. I stopped short of giving a worked example -- maybe later.

popular enough, Flux maintainers can add them to the Flux image
(please create an issue). In the future it may be possibly to specify
an container image for each command.

The working directory (also known as CWD) of the `command`s executed
from a `.flux.yaml` file will be set to the target path, i.e., the
`--git-path` entry.

For example, when using flux with `--git-path=staging` on a git
repository with this structure:

```sh
├── .flux.yaml
├── staging/
├──── [...]
├── production/
└──── [...]
```

The commands in `.flux.yaml` will be executed with their working
directory set to `staging`.

> In addition, `updaters` are given arguments via environment variables:

Reviewer: Nitpick:

```suggestion
In addition, the `updaters` of the `commandUpdated` configuration are given arguments via environment variables:
```

Reviewer 2: I would include an example of `kustomize edit set image`

Author: I did not attempt to give a working example of `commandUpdated` updaters, because the thought of trying to formulate one terrified me. `kustomize edit set imagetag` sets the image for _all things that use it_, last time I looked.

# Manifest generation through `.flux.yaml` configuration files

This feature lets you generate Kubernetes manifests with a program,
instead of having to include them in your git repo as YAML files. For
example, you can use `kustomize` to patch a common set of resources to
suit a particular environment.

Manifest generation is controlled by the flags given to `fluxd`, and
`.flux.yaml` files in your git repo.

To enable it, you will need to

 1. pass the command-line flag `--manifest-generation=true`
to `fluxd`.
 2. put at least one `.flux.yaml` file in the git repository.

Where to put `.flux.yaml`, and what should be in it, are described in
the sections following.

## Setting manifest generation up

The command-line flag `--git-path` (which can be given multiple
values) marks a "target path" within the git repository in which to

> find manifests. If `--git-path` is not supplied, the top of the git

Reviewer: Maybe it's worth indicating that multiple `--git-paths` can be provided?

Author: You mean like saying "which can be given multiple values"?

Reviewer: Yes. Note that it's also valid to provide multiple `--git-path` arguments (pflag will coalesce them)

Author: Yup, that's what I mean by "provide multiple values" -- use the flag to provide multiple values. I don't want to get into how the flag parsing works in this doc though.

│   ├── podinfo-hpa.yaml
│   └── podinfo-svc.yaml
├── .flux.yaml
├── production
│   ├── flux-patch.yaml
│   ├── kustomization.yaml
│   └── replicas-patch.yaml
└── staging
    ├── flux-patch.yaml
    └── kustomization.yaml
```

In this case, say you started `fluxd` with `--git-path=staging`, it
would find `.flux.yaml` in the top directory and use that to generate
manifests.  The other files and directories (if there were any) in
`staging/` are not examined by fluxd, in favour of following the
instructions given in the `.flux.yaml` file.

This layout could also be used with `--git-path=production`.

In this modified example, the `.flux.yaml` file has been moved under
`staging/`:

```
.
├── base
│   ├── demo-ns.yaml
│   ├── kustomization.yaml
│   ├── podinfo-dep.yaml
│   ├── podinfo-hpa.yaml
│   └── podinfo-svc.yaml
├── production
│   ├── flux-patch.yaml
│   ├── kustomization.yaml
│   └── replicas-patch.yaml
└── staging
    ├── flux-patch.yaml
    ├── .flux.yaml
    └── kustomization.yaml
```

… since the `.flux.yaml` file is now under `staging/`, it will still
take effect for `--git-path=staging`. However:

Using `--git-path=production` would **not produce a usable

> configuration**, because without an applicable `.flux.yaml`, the files

Reviewer: Even if redundant, I think it's worth indicating that a `.flux.yaml` file cannot be found in that case because there is non in `/production` or its parent directories (`/`)

stdout. The streams are concatenated and parsed as one big YAML
stream, before being applied.

Much of the time, it will only be necessary to supply one command to
be run.

The commands will be run with the target path being processed as a
working directory -- which is not necessarily the same directory in
which the `.flux.yaml` file was found. [See below](#execution-context)
for more details on the execution context in which commands are run.

### Using patch-updated configuration

A patch-updated configuration generates manifests using commands, and
records updates as a set of [strategic merge][strategic-merge] patches
in a file.

For example, when an automated image upgrade is run, fluxd will do
this:

 1. run the generator commands and parse the manifests;
 2. find the manifest that needs to be updated, and calculate the
    patch to it that performs the update;
 3. record that patch in the patch file.

When syncing, fluxd will generate the manifests as usual, then apply
all the patches that have been recorded in the patch file.

This is how a patch-updated `.flux.yaml` looks:

```yaml
version: 1
patchUpdated:
  generators:

>   - command: jk generate ./index.js

Reviewer: It may be worth indicating that `jk` is not in the container, or indicate how to install it, or simply use `kustomize`. Otherwise I think it may be confusing.

Author: I suppooooooose

Author: Changed it to be `generator_command`, so as not to distract from the shape of the YAML, which is the point here.
